cmd/docker/docker.go

```diff
@@ -28,16 +28,43 @@ import (
 	"go.opentelemetry.io/otel"
 )
 
+var errCtxUserTerminated = errors.New("user terminated the process")
+
 func main() {
-	err := dockerMain(context.Background())
-	if err != nil && !errdefs.IsCancelled(err) {
+	ctx := context.Background()
+	err := dockerMain(ctx)
+	if err != nil && !errdefs.IsCancelled(err) && !errors.Is(err, errCtxUserTerminated) {
 		_, _ = fmt.Fprintln(os.Stderr, err)
 		os.Exit(getExitCode(err))
 	}
 }
 
+func notifyContext(ctx context.Context, signals ...os.Signal) (context.Context, context.CancelFunc) {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, signals...)
+
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			signal.Stop(ch)
+			return
+		case <-ch:
+			cancel(errCtxUserTerminated)
```

Comment:
The thing that I can't wrap my mind around is:

After this cancel, we'll have:
```
ctx.Err() = context.Canceled
context.Cause(ctx) = errCtxUserTerminated
```

With that in mind, how `dockerMain` ends up returning the cause, instead of the `ctx.Err()`? 🤔

Reply:
Yes you are right, essentially right now not all commands would return a cause, but the standard library _should_ - like http.NewRequestWithContext.

Reply:
Go std `http.Transport` returns a `context.Cause` when `context.Done()`
https://github.com/golang/go/blob/608acff8479640b00c85371d91280b64f5ec9594/src/net/http/transport.go#L669